src/dsp_tools/utils/xml_upload_stash.py

```diff
@@ -179,13 +198,18 @@ def upload_single_link_xml_property(
 
     # if the pure text is a hash, the replacement must be made
     # this hash originates from _stash_circular_references(), and identifies the XML texts
-    if text_hash_value not in hash_to_value:
+    try:
+        xml_from_stash = hash_to_value[text_hash_value]
+    except KeyError as err:
+        _log_iri_does_not_exist_error(received_error=err, stashed_resource=stashed_resource, all_link_props=link_prop)
         # no action necessary: this property will remain in nonapplied_xml_texts,
         # which will be handled by the caller
         return nonapplied_xml_texts
 
-    new_xmltext = _replace_internal_ids_with_iris(
-        pure_text=text_hash_value, id2iri_mapping=id2iri_mapping, hash_to_value=hash_to_value
+    id_set = xml_from_stash.find_all_substring_in_xmlstr(pattern='href="IRI:(.*?):IRI"')
```

Comment:
this is a pattern that always remains the same and is part of the "domain logic" of the application, so it might be worth extracting it into a top level constant